### ISSUE TITLE
Union type assertions shouldn't care about the order

### ISSUE THREAD

**Issue by benlesh** [Microsoft/dtslint#191]
We've hit an issue in the [RxJS dtslint tests](https://github.com/ReactiveX/rxjs/pull/4516/files#diff-18fc411fb62cc3e31431e1f5a691e306L39) where a test is trying to assert:

```ts
const o = from([of(1), ['test'], iterable]); // $ExpectType Observable<IterableIterator<number> | Observable<number> | string[]>
```

And dtslint complains it wants `Observable<Observable<number> | IterableIterator<number> | string[]>`, but if you change it to that:

```ts
const o = from([of(1), ['test'], iterable]); // $ExpectType Observable<Observable<number> | IterableIterator<number> | string[]>
```

... then it complains that you should change it back.

It seems like this is the wrong behavior:  `a | b | c` is no different than `c | b | a`, AFAICT. ... but perhaps I'm wrong?

Thank you in advance for your help.

**Comment by SgLy** [comment]
I am having the same problem. dtslint reports that `$ExpectType 1 | 2` but got `2 | 1` and gives me an error.

**Comment by ersimont** [comment]
I am hitting this issue. When I make certain changes to me production `.ts` files, `dtslint` suddenly changes the expected order of many union types in unrelated files, causing many of their typing tests to fail.

Either order should not matter for unions (as the title of this issue suggests), or the order should remain stable. As it is, I can't figure out how to add the new function I want to my library without updating ~70 type expectations. And then what if the next function I add swaps them back?

**Comment by ersimont** [comment]
I made a branch on [`micro-dash`](https://github.com/simontonsoftware/micro-dash) to demonstrate the issue. To reproduce:

1. Check out the branch [`dts-union-swap`](https://github.com/simontonsoftware/micro-dash/compare/dts-union-swap)
1. `yarn install`
1. `yarn dtslint` - everything passes
1. Uncomment the contents of `projects/micro-dash/src/typing-tests/at.dts-spec.ts`
1. `yarn dtslint` - many errors because the ordering of unions changed

**Comment by thw0rted** [comment]
There seems to be a large class of errors that stem from the fact that, while `expectType` (from `tsd`) takes an actual "type-space" generic param as the expectation, the dtslint `$ExpectType` does simple string matching:

https://github.com/microsoft/dtslint/issues/57
https://github.com/microsoft/dtslint/issues/114
https://github.com/microsoft/dtslint/issues/352

Would it be better to just transition to using `tsd` tests instead of the comment-based syntax?  (I'm not that familiar with the ecosystem, I assume the dtslint implementation has some kind of advantages over `tsd`...)

**Comment by thw0rted** [comment]
Very interesting: I just found https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60016.  Until that edit, I didn't realize that an or-operator (`||`) worked in `$ExpectType` -- if it's documented anywhere, I couldn't find it.  So, I guess one workaround for this issue, if you have a small union, would be `$ExpectType A | B || B | C`.  (In that PR, it's used to handle different versions of TS ordering the union in question differently.)
